exporters/etw/include/opentelemetry/exporters/etw/etw_provider_exporter.h

```diff
@@ -0,0 +1,664 @@
+        providerMetaBuilder.AddTrait(tld::ProviderTraitType::ProviderTraitGroupGuid,
+                                     (void *)&providerGroupGuid, sizeof(GUID));
+        providerMetaBuilder.End();
+
+        REGHANDLE hProvider = 0;
+        if (0 !=
+            tld::RegisterProvider(&hProvider, &data.providerGuid, data.providerMetaVector.data()))
+        {
+          // There was an error registering the ETW provider
+          data.providerHandle = INVALID_HANDLE;
+        }
+        else
+        {
+          data.providerHandle = hProvider;
+        };
+      };
+      break;
+#endif
+
+#ifdef HAVE_MSGPACK
+      // Register for MsgPack payload ETW events.
+      case EventFormat::ETW_MSGPACK:
+      {
+        REGHANDLE hProvider = 0;
+        if (EventRegister(&data.providerGuid, NULL, NULL, &hProvider) != ERROR_SUCCESS)
+        {
+          // There was an error registering the ETW provider
+          data.providerHandle = INVALID_HANDLE;
+        }
+        else
+        {
+          data.providerHandle = hProvider;
+        }
+      };
+      break;
+#endif
+
+      default:
+        // TODO: other protocols, e.g. XML events - not supported yet
+        break;
+    }
+
+    // We always return an entry even if we failed to register.
+    // Caller should check whether the hProvider handle is valid.
+    return data;
+  }
+
+  /// <summary>
+  /// Unregister Provider
+  /// </summary>
+  /// <param name="providerId"></param>
+  /// <returns></returns>
+  unsigned long close(Handle data)
+  {
+    std::lock_guard<std::mutex> lock(m_providerMapLock);
+
+    auto m  = providers();
+    auto it = m.begin();
+    while (it != m.end())
+    {
+      if (it->second.providerHandle == data.providerHandle)
+      {
+        auto result = EventUnregister(data.providerHandle);
+        m.erase(it);
+        return result;
+      }
+    };
+    return STATUS_ERROR;
+  }
+
+#ifdef HAVE_MSGPACK
+  template <class T>
+  unsigned long writeMsgPack(Handle &providerData, T eventData)
+  {
+
+    // Make sure you stop sending event before register unregistering providerData
+    if (providerData.providerHandle == INVALID_HANDLE)
+    {
+      // Provider not registered!
+      return STATUS_ERROR;
+    };
+
+    const std::string EVENT_NAME = "name";
+    std::string eventName        = "NoName";
+    auto nameField               = eventData[EVENT_NAME];
+    switch (nameField.index())
+    {
+      case common::AttributeType::TYPE_STRING:
+        eventName =
+            (char *)(nostd::get<nostd::string_view>(nameField).data());  // must be 0-terminated!
+        break;
+      case common::AttributeType::TYPE_CSTRING:
+        eventName = (char *)(nostd::get<const char *>(nameField));
+        break;
+      default:
+        // This is user error. Invalid event name!
+        // We supply default 'NoName' event name in this case.
+        break;
+    }
+
+    /* clang-format off */
+    nlohmann::json jObj =
+    {
+      { "env_name", "Span" },
```

Comment:
Let's discuss what attribute names we use here, because `env_name` and `env_ver` were originally from another implementation that may use some Microsoft-specific schema.

Reply:
Tracking it https://github.com/open-telemetry/opentelemetry-cpp/issues/466 - which **I think we should solve in a separate PR**.

Reply:
This will be covered in a follow-up PR.